packages/wrangler/src/api/startDevWorker/BundlerController.ts

```diff
@@ -1,37 +1,280 @@
+				const destinationDir = this.#tmpDir.path;
+				writeFileSync(
+					path.join(destinationDir, path.basename(config._entry.file)),
+					readFileSync(config._entry.file, "utf-8")
+				);
+			}
+
+			const entryDirectory = path.dirname(config._entry.file);
+			const moduleCollector = createModuleCollector({
+				wrangler1xLegacyModuleReferences: getWrangler1xLegacyModuleReferences(
+					entryDirectory,
+					config._entry.file
+				),
+				entry: config._entry,
+				// `moduleCollector` doesn't get used when `props.noBundle` is set, so
+				// `findAdditionalModules` always defaults to `false`
+				findAdditionalModules: config.build.findAdditionalModules ?? false,
+				rules: config.build.moduleRules,
+			});
+
+			const bundleResult: Omit<BundleResult, "stop"> = !config.build?.bundle
+				? await noBundleWorker(
+						config._entry,
+						config.build.moduleRules,
+						this.#tmpDir.path
+					)
+				: await bundleWorker(config._entry, this.#tmpDir.path, {
+						bundle: true,
+						additionalModules: [],
+						moduleCollector,
+						serveAssetsFromWorker: Boolean(config._serveAssetsFromWorker),
+						doBindings: config._bindings?.durable_objects?.bindings ?? [],
+						jsxFactory: config.build.jsxFactory,
+						jsxFragment: config.build.jsxFactory,
+						tsconfig: config.build.tsconfig,
+						minify: config.build.minify,
+						nodejsCompatMode: config.build.nodejsCompatMode,
+						define: config.build.define,
+						checkFetch: true,
+						assets: config._assets,
+						// enable the cache when publishing
+						bypassAssetCache: false,
+						// We want to know if the build is for development or publishing
+						// This could potentially cause issues as we no longer have identical behaviour between dev and deploy?
+						targetConsumer: "dev",
+						local: !config.dev?.remote,
+						projectRoot: config._projectRoot,
+						defineNavigatorUserAgent: isNavigatorDefined(
+							config.compatibilityDate,
+							config.compatibilityFlags
+						),
+					});
+			if (buildAborter.signal.aborted) {
+				return;
+			}
+			const entrypointPath = realpathSync(
+				bundleResult?.resolvedEntryPointPath ?? config._entry.file
+			);
+			this.emitBundleCompletedEvent({
+				type: "bundleComplete",
+				config,
+				bundle: {
+					id: 0,
+					entry: config._entry,
+					path: entrypointPath,
+					type:
+						bundleResult?.bundleType ??
+						getBundleType(config._entry.format, config._entry.file),
+					modules: bundleResult.modules,
+					dependencies: bundleResult?.dependencies ?? {},
+					sourceMapPath: bundleResult?.sourceMapPath,
+					sourceMapMetadata: bundleResult?.sourceMapMetadata,
+					entrypointSource: readFileSync(entrypointPath, "utf8"),
+				},
+			});
+		} catch (err) {
+			logger.error("Custom build failed:", err);
+			this.emitErrorEvent({
+				type: "error",
+				reason: "Custom build failed",
+				cause: castErrorCause(err),
+				source: "BundlerController",
+				data: undefined,
```

Comment:
I think I expected this to be used for the inputs that caused the step to run – on error it will help dry-run the function if the inputs are known

```suggestion
				data: { config, filePath },
```